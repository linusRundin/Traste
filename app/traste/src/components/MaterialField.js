import React from 'react';
import PropTypes from 'prop-types';
import {Colors} from '../assets/Colors';

import {TextField, InputAdornment} from '@mui/material';

/**
 * Custom TextField for displaying a procentage for a specified material.
 * @param {*} label Name of field.
 * @param {*} value Starting value.
 * @param {*} onChange Handler.
 * @param {*} error What will be shown for the error.
 * @return {TextField} Rendred field.
 */
function MaterialField({label, value, onChange, error}) {
  return (
    <TextField
      fullWidth
      label={label}
      InputProps={{
        endAdornment: (
          <InputAdornment position='start' disableTypography={true}>
            %
          </InputAdornment>
        ),
      }}
      variant='outlined'
      type='number'
      onFocus={(event) => {
        event.target.select();
      }}
      onBlur={(evt) => {
        if (evt.target.value === '' || evt.target.value < 0) {
          onChange(0);
        }
      }}
      value={value}
      onChange={(e) => {
        let tmpval = e.target.value;
        if (isNaN(parseInt(e.target.value, 10))) {
          tmpval = '';
        } else {
          tmpval = parseInt(tmpval, 10);
        }
        onChange(tmpval);
      }}
      sx={{
        backgroundColor: Colors.transparentWhite,
      }}
      inputProps={{
        'style': {textAlign: 'right', fontSize: 16},
        'data-testid': 'materialfield',
      }}
      error={!!error}
    />
  );
}

MaterialField.defaultProps = {
  error: false,
};

MaterialField.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.any.isRequired,
};

export default MaterialField;
