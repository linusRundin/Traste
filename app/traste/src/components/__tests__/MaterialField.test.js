import {render, screen, cleanup, fireEvent} from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import MaterialField from '../MaterialField';
import React from 'react';

afterEach(() => {
  cleanup();
});

test('Should render MaterialField with label Wood', () => {
  const propsData = {label: 'Wood',
    value: 'string', error: 'error'};
  const onChange = jest.fn();
  render(<MaterialField {...propsData} onChange={onChange}/>);

  const materialFieldElement = screen.getByTestId('materialfield');
  expect(materialFieldElement).toBeInTheDocument();
  expect(materialFieldElement).toHaveTextContent('');
  expect(materialFieldElement).toHaveAccessibleName('Wood', 'Wood');
});

test('Should not render MaterialField with empty label', () => {
  const propsData = {label: 'Wood', value: 'string', error: 'error'};
  const onChange = jest.fn();
  render(<MaterialField {...propsData} onChange={onChange} />);

  const materialFieldElement = screen.getByTestId('materialfield');
  expect(materialFieldElement).toBeInTheDocument();
  expect(materialFieldElement).not.toHaveAccessibleName('', 'Wood');
});

test('Test MaterialField onChange', () => {
  const propsData = {label: 'Wood', error: 'error', value: 'tes'};
  const onChange = jest.fn();
  render(<MaterialField {...propsData} onChange={onChange} />);

  const materialFieldElement = screen.getByTestId('materialfield');

  // Test what happens when you input a number.
  fireEvent.change(materialFieldElement, {target: {value: '10'}});
  expect(onChange).toHaveBeenCalledTimes(1);

  // Test what happens when you input something that is not a number.
  fireEvent.change(materialFieldElement, {target: {value: 'apa'}});
  expect(onChange).toHaveBeenCalledTimes(2);
});

test('Test MaterialField onFocus', () => {
  const propsData = {label: 'Wood', error: 'error', value: 'tes'};
  const onChange = jest.fn();
  render(<MaterialField {...propsData} onChange={onChange}/>);
  const materialFieldElement = screen.getByTestId('materialfield');

  materialFieldElement.focus();
  expect(materialFieldElement).toHaveFocus();
});

test('Test MaterialField onBlur', () => {
  const propsData = {label: 'Wood', error: 'error', value: 'tes'};
  const onChange = jest.fn();
  render(<MaterialField {...propsData} onChange={onChange} />);
  const materialFieldElement = screen.getByTestId('materialfield');

  fireEvent.blur(materialFieldElement, {target: {value: 0}});
  fireEvent.blur(materialFieldElement, {target: {value: -1}});
});

test('Matches snapshot', () => {
  const propsData = {label: 'Wood', value: 'string', error: 'error'};
  const onChange = jest.fn();

  const tree = renderer.create(<MaterialField {...propsData}
    onChange={onChange} />).toJSON();
  expect(tree).toMatchSnapshot();
});
