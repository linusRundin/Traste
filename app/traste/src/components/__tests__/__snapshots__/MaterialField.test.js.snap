// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-fhe0il-MuiFormControl-root-MuiTextField-root"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
  >
    Wood
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={true}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
      data-testid="materialfield"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      style={
        Object {
          "fontSize": 16,
          "textAlign": "right",
        }
      }
      type="number"
      value="string"
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
    >
      <span
        className="notranslate"
      >
        â€‹
      </span>
      %
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1z7n62"
      >
        <span>
          Wood
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
